adapter/service/virtualmeter/virtual_meter_manager.go

```diff
@@ -0,0 +1,332 @@
+	if oldModes != nil {
+		if err := m.recalculateEnergy(&device); err != nil {
+			return fmt.Errorf("failed to update energy: %w", err)
+		}
+	}
+
+	device.Modes = modes
+	device.Unit = unit
+
+	if err := m.storage.SetDeviceEntry(addr, device); err != nil {
+		return fmt.Errorf("failed add meter, can't save data: %w", err)
+	}
+
+	// Update thing only if a service has been just added.
+	if oldModes == nil {
+		if err := thing.Update(true, adapter.ThingUpdateAddService(s)); err != nil {
+			return fmt.Errorf("failed to update thing. Can't add service. Addr: %s. %w", addr, err)
+		}
+	}
+
+	return nil
+}
+
+func (m *virtualMeterManager) Remove(addr string) error {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	addr = m.thingAddrFromTopic(addr)
+
+	s := m.serviceTemplates[addr]
+	if s == nil {
+		return fmt.Errorf("failed to remove meter from a thing: %s. No service template found", addr)
+	}
+
+	thing := m.ad.ThingByAddress(addr)
+	if thing == nil {
+		return fmt.Errorf("no thing found by address: %s. can't remove meter", addr)
+	}
+
+	if err := m.storage.DeleteDevice(addr); err != nil {
+		return fmt.Errorf("failed to delete meter, can't remove from storage: %w", err)
+	}
+
+	if err := thing.Update(true, adapter.ThingUpdateRemoveService(m.serviceTemplates[addr])); err != nil {
+		return fmt.Errorf("failed to update thing. Can't remove service. Addr: %s. %w", addr, err)
+	}
+
+	return nil
+}
+
+func (m *virtualMeterManager) Modes(addr string) (map[string]float64, error) {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+
+	addr = m.thingAddrFromTopic(addr)
+
+	device, err := m.storage.Device(addr)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get modes: %w", err)
+	}
+
+	return device.Modes, nil
+}
+
+func (m *virtualMeterManager) SetReportingInterval(duration time.Duration) error {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	if m.taskManager == nil {
+		return fmt.Errorf("virtual meter does't have task manager initialised, can't update task interval")
+	}
+
+	if err := m.storage.SetReportingInterval(duration); err != nil {
+		return fmt.Errorf("failed to save repoting interval: %w", err)
+	}
+
+	if err := m.taskManager.UpdateTaskInterval(TaskVirtualReporter, duration); err != nil {
+		return fmt.Errorf("failed to update a task with new reporting interval: %w", err)
+	}
+
+	return nil
+}
+
+func (m *virtualMeterManager) ReportingInterval() time.Duration {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+
+	return m.storage.ReportingInterval()
+}
+
+func (m *virtualMeterManager) RegisterDevice(thing adapter.Thing, addr string, service adapter.Service) error {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	addr = m.thingAddrFromTopic(addr)
+
+	log.Infof("Virtual meter: registering a service template, addr: %s", addr)
+
+	device, err := m.storage.Device(addr)
+	if err != nil && !errors.As(err, &ErrorEntryNotFound{}) {
+		return fmt.Errorf("virtual meter: failed to get device by addr %s: %w", addr, err)
+	} else if err == nil && device.Modes != nil {
```

Comment:
In logic like this rather check length, because empty initialized map/slice would pass.